File: LeetCode/Python/Parsing A Boolean Expression.py
"""
  Parsing A Boolean Expression
  URL: https://leetcode.com/problems/parsing-a-boolean-expression
  Difficulty: Hard
  Topic : String, Stack, Recursion
"""

from functools import reduce
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        def operate(o, v: list[bool]) -> bool:
            if o == "&":
                return reduce(lambda x, r: x and r, v, True)
            elif o == "|":
                return reduce(lambda x, r: x or r, v, False)
            else:
                return not v[0]
        
        operStack = []
        boolStack: list[bool] = []
        for e in expression:
            if e in "&|!":
                operStack.append(e)
            elif e == "(":
                boolStack.append([])
            elif e == ")":
                r = operate(operStack.pop(), boolStack.pop())
                print(r)
                if len(boolStack) > 0:
                    last = boolStack.pop()
                    last.append(r)
                    boolStack.append(last)
                else:
                    boolStack.append([r])
            elif e in "ft":
                if len(boolStack) > 0:
                    last = boolStack.pop()
                    last.append(e == "t")
                    boolStack.append(last)
        return boolStack[-1][-1]

File: LeetCode/Hard/Parsing A Boolean Expression.swift
/*
  Parsing A Boolean Expression
  URL: https://leetcode.com/problems/parsing-a-boolean-expression
  Difficulty: Hard
  Topic : String, Stack, Recursion
*/

class Solution {
    func parseBoolExpr(_ expression: String) -> Bool {
        func operate(o: String, v: [Bool]) -> Bool {
            if o == "&" {
                return v.reduce(true) { $0 && $1 }
            } else if o == "|" {
                return v.reduce(false) { $0 || $1 }
            } else {
                return !v.first!
            }
        }
        var operStack: [String] = []
        var boolStack: [[Bool]] = []
        for e in expression {
            if "\(e)" == "&" || "\(e)" == "|" || "\(e)" == "!" {
                operStack.append("\(e)")
            } else if "\(e)" == "(" {
                boolStack.append([])
            } else if "\(e)" == ")" {
                let oper = operStack.popLast()!
                let value = boolStack.popLast()!
                let r = operate(o: oper, v: value)
                if var last = boolStack.popLast() {
                    last.append(r)
                    boolStack.append(last)
                } else {
                    boolStack.append([r])
                }
            } else if "\(e)" == "f" || "\(e)" == "t" {
                if var last = boolStack.popLast() {
                    last.append("\(e)" == "t" ? true : false)
                    boolStack.append(last)
                }
            }
        }
        return boolStack.last?.last ?? true
    }
}
